controller/internal/enforcer/envoyauthorizer/envoyauthorizerenforcer.go

```diff
@@ -82,16 +101,201 @@ func (e *Enforcer) Secrets() (secrets.Secrets, func()) {
+
+	// now instantiate the apiAuth and metadata
+	// create a new server if it doesn't exist yet
+	//return nil
+	if _, err := e.clients.Get(contextID); err != nil {
+		zap.L().Debug("creating new auth and sds servers", zap.String("puID", contextID))
+		ingressServer, err := envoyproxy.NewExtAuthzServer(contextID, e.puContexts, e.collector, envoyproxy.IngressDirection, e.registry, e.secrets, e.tokenIssuer)
+		if err != nil {
+			zap.L().Error("Cannot create and run IngressServer", zap.Error(err))
+			return err
+		}
+
+		egressServer, err := envoyproxy.NewExtAuthzServer(contextID, e.puContexts, e.collector, envoyproxy.EgressDirection, e.registry, e.secrets, e.tokenIssuer)
+		if err != nil {
+			zap.L().Error("Cannot create and run EgressServer", zap.Error(err))
+			ingressServer.Stop()
+			return err
+		}
+		sdsServer, err := envoyproxy.NewSdsServer(contextID, puInfo, caPool, e.secrets)
+		if err != nil {
+			zap.L().Error("Cannot create and run SdsServer", zap.Error(err))
+			return err
+		}
+		//sdsServer := &envoyproxy.SdsServer{}
+		// Add the EnvoyServers to our cache
+		if err := e.clients.Add(contextID, &envoyServers{ingress: ingressServer, egress: egressServer, sds: sdsServer}); err != nil {
+			ingressServer.Stop()
+			egressServer.Stop()
+			sdsServer.Stop()
+			return err
+		}
+
+	} else {
+		// we have this client already, this is only a policy update
+		zap.L().Debug("handling policy update for envoy servers", zap.String("puID", contextID))
+		// For updates we need to update the certificates if we have new ones. Otherwise
+		// we return. There is nothing else to do in case of policy update.
+		// this required for the Envoy sds. So that the SDS picks the latest cert.
+		if c, cerr := e.clients.Get(contextID); cerr == nil {
+			_, perr := e.processCertificateUpdates(puInfo, c.(*envoyServers).sds, caPool)
+			if perr != nil {
+				zap.L().Error("unable to update certificates for services", zap.Error(perr))
+				return perr
+			}
+			return nil
+		}
+	}
+
 	return nil
 }
 
+// processCertificateUpdates processes the certificate information and updates
+// the servers.
+func (e *Enforcer) processCertificateUpdates(puInfo *policy.PUInfo, server *envoyproxy.SdsServer, caPool *x509.CertPool) (bool, error) {
+
+	// If there are certificates provided, we will need to update them for the
+	// services. If the certificates are nil, we ignore them.
+	certPEM, keyPEM, caPEM := puInfo.Policy.ServiceCertificates()
+	if certPEM == "" || keyPEM == "" {
+		return false, nil
+	}
+
+	// Process any updates on the cert pool
+	if caPEM != "" {
+		if !caPool.AppendCertsFromPEM([]byte(caPEM)) {
+			zap.L().Warn("Failed to add Services CA")
+		}
+	}
+
+	// Create the TLS certificate
+	tlsCert, err := tls.X509KeyPair([]byte(certPEM), []byte(keyPEM))
+	if err != nil {
+		return false, fmt.Errorf("Invalid certificates: %s", err)
+	}
+	// update the sds server certs.
+	server.UpdateSecrets(&tlsCert, caPool, e.secrets, certPEM, keyPEM)
```

Comment:
according to our docs, remote enforcers are also getting their `e.secrets` updated only through a policy update (haven't fully followed that code path yet)

anyway, same as for the auth servers, caution when accessing/updating the secrets ... that's exactly why I added the `LockedSecrets` interface to this enforcer, for a single point of access to the secrets